### PR TITLE
CameraView iOS

### PR DESCRIPTION
**Pull request by jfversluis**
### Description of Change ###

Adds the iOS part (and shared part) of the `CameraView`

### Bugs Fixed ###

- Related to issue #211 

### API Changes ###

Added: 

- `public enum CameraOptions` enum for selecting front, back, default or external camera
- `public enum CameraFlashMode` enum for enabling, disabling, torch or auto the flash
- `public enum CameraCaptureOptions` enum for setting the camera to photo, video or default

```csharp
public class CameraMediaSource : MediaSource
{
	public static MediaSource FromStream(Func<Stream> stream) ;
	public static MediaSource FromResource(string resource, Type resolvingType);
	public static MediaSource FromResource(string resource, Assembly sourceAssembly = null);
}
```

```csharp
public class StreamMediaSource : MediaSource, IStreamImageSource
{
	protected CancellationTokenSource CancellationTokenSource { get; }
	public virtual Func<CancellationToken, Task<Stream>> Stream { get; set; } // Bindable
	protected override void OnPropertyChanged(string propertyName);
	protected void OnLoadingCompleted(bool cancelled);
	protected void OnLoadingStarted();
	public virtual Task<bool> Cancel();
}
```

```csharp
public class MediaCapturedEventArgs : EventArgs
{
         internal MediaCapturedEventArgs();
	public object Data { get; }
	public ImageSource Image { get; }
	public MediaSource Video { get; }
}
```

```csharp
public class CameraView : View
{
	public event EventHandler<MediaCapturedEventArgs> MediaCaptured;
	public event EventHandler<string> MediaCaptureFailed;
	public event EventHandler<bool> OnAvailable;

	public bool IsBusy { get; set; } // Bindable
	public bool IsAvailable { get; set; } // Bindable
	public CameraOptions CameraOptions { get; set; } // Bindable
	public bool SavePhotoToFile { get; set; } // Bindable
	public CameraCaptureOptions CaptureOptions { get; set; } // Bindable
	public bool KeepScreenOn { get; set; } // Bindable
	public bool VideoStabilization { get; set; } // Bindable
	public CameraFlashMode FlashMode { get; set; } // Bindable
	public Aspect PreviewAspect { get; set; } // Bindable
	public double Zoom { get; set; } // Bindable
	public double MaxZoom  { get; set; } // Bindable
	public void Shutter();

         internal event EventHandler ShutterClicked;
	internal void RaiseMediaCaptured(MediaCapturedEventArgs args);
	internal void RaiseMediaCaptureFailed(string message);
} 
```

### Behavioral Changes ###

New control, so should have any behavioral changes :)

### PR Checklist ###

- [x] Has tests (if omitted, state reason in description)
- [x] Has samples (if omitted, state reason in description)
- [x] Rebased on top of main at time of PR
- [x] Changes adhere to coding standard
- [ ] Updated documentation

### CODE REVIEW

File: README.md
Comment (jfversluis) [inline README.md:67]: Rearranged this for a bit. Added `Core` and made the rest more clear

File: XamarinCommunityToolkit/Core/CameraMediaSource.shared.cs
Comment (jfversluis) [inline XamarinCommunityToolkit/Core/CameraMediaSource.shared.cs:11]: If/when `MediaElement` is dropped from Forms, we might also get the entire `MediaSource` in the Toolkit. Until that happens like this? Is this used in Android or UWP? It doesn't seem to be used in iOS.

File: XamarinCommunityToolkit/Extensions/TranslateExtension.shared.cs
Comment (jfversluis) [inline XamarinCommunityToolkit/Extensions/TranslateExtension.shared.cs:27]: Not really related, but added this to make it a bit more clear to the end user

File: XamarinCommunityToolkitSample/ViewModels/Views/ViewsGalleryViewModel.cs
Comment (jfversluis) [inline XamarinCommunityToolkitSample/ViewModels/Views/ViewsGalleryViewModel.cs:14]: Removed the color specifications here, that is arranged automagically

File: XamarinCommunityToolkit/Views/CameraView/CameraView.shared.cs
Comment (AndreiMisiukevich) [inline XamarinCommunityToolkit/Views/CameraView/CameraView.shared.cs:97]: Would you like to replace useless "float" by super-duper "double" ?)

File: XamarinCommunityToolkit/Core/CameraMediaSource.shared.cs
Comment (AndreiMisiukevich) [inline XamarinCommunityToolkit/Core/CameraMediaSource.shared.cs:18]:
```suggestion
		=> FromResource(resource, resolvingType.GetTypeInfo().Assembly);
```

File: XamarinCommunityToolkit/Core/CameraMediaSource.shared.cs
Comment (AndreiMisiukevich) [inline XamarinCommunityToolkit/Core/CameraMediaSource.shared.cs:13]:
```suggestion
		public static MediaSource FromStream(Func<Stream> stream) 
			=> new StreamMediaSource { Stream = token => Task.Run(stream, token) };
```

File: XamarinCommunityToolkit/Core/CameraMediaSource.shared.cs
Comment (AndreiMisiukevich) [inline XamarinCommunityToolkit/Core/CameraMediaSource.shared.cs:36]:
```suggestion
				if (callingAssemblyMethod == null)
				{
					Forms.Internals.Log.Warning("Warning", "Can not find CallingAssembly, pass resolvingType to FromResource to ensure proper resolution");
					return null;
				}
				sourceAssembly = (Assembly)callingAssemblyMethod.Invoke(null, new object[0]);
```

File: XamarinCommunityToolkit/Core/StreamMediaSource.shared.cs
Comment (AndreiMisiukevich) [inline XamarinCommunityToolkit/Core/StreamMediaSource.shared.cs:17]:
```suggestion
		public static readonly BindableProperty StreamProperty 
			= BindableProperty.Create(nameof(Stream), typeof(Func<CancellationToken, Task<Stream>>), typeof(StreamMediaSource));
```

Comment (AndreiMisiukevich) [reply XamarinCommunityToolkit/Core/StreamMediaSource.shared.cs:17]: We can skip passing default(Func<CancellationToken, Task<Stream>>) cuz it's null. Null is a default value for "defaultValue" parameter.

File: XamarinCommunityToolkit/Views/CameraView/CameraView.shared.cs
Comment (AndreiMisiukevich) [inline XamarinCommunityToolkit/Views/CameraView/CameraView.shared.cs:12]:
```suggestion
```

**Review by AndreiMisiukevich — APPROVED** [review]
I don't want to block this PR from being merged.

File: XamarinCommunityToolkit/Core/StreamMediaSource.shared.cs
Comment (AndreiMisiukevich) [inline XamarinCommunityToolkit/Core/StreamMediaSource.shared.cs:50]: Do we need to implement this interface Explicitly?

File: XamarinCommunityToolkit/Core/StreamMediaSource.shared.cs
Comment (AndreiMisiukevich) [inline XamarinCommunityToolkit/Core/StreamMediaSource.shared.cs:79]: I guess this lock should be moved inside of CancellationTokenSource setter, no?

File: XamarinCommunityToolkit/Core/StreamMediaSource.shared.cs
Comment (AndreiMisiukevich) [inline XamarinCommunityToolkit/Core/StreamMediaSource.shared.cs:108]:
```suggestion
			if (original == null)
				CancellationTokenSource = null;
			else
				tcs = original;
```

File: XamarinCommunityToolkit/Core/StreamMediaSource.shared.cs
Comment (jfversluis) [reply XamarinCommunityToolkit/Core/StreamMediaSource.shared.cs:50]: It would appear so. I'm not convinced that we need it at all, but let's see if we can clean it up when we get the other platforms in here

File: XamarinCommunityToolkit/Core/CameraMediaSource.shared.cs
Comment (pictos) [reply XamarinCommunityToolkit/Core/CameraMediaSource.shared.cs:11]: I think this is just to play the VideoRecord.

Comment (pictos) [reply XamarinCommunityToolkit/Core/CameraMediaSource.shared.cs:11]: It's used to retrieve the video information. Is used [here](https://github.com/xamarin/XamarinCommunityToolkit/pull/244/files#diff-61ca214876a712c1f7a458c977572116R10)

File: XamarinCommunityToolkit/Core/StreamMediaSource.shared.cs
Comment (pictos) [reply XamarinCommunityToolkit/Core/StreamMediaSource.shared.cs:79]: @AndreiMisiukevich, do you have a code snippet for that?

File: XamarinCommunityToolkit/Views/CameraView/CameraView.shared.cs
Comment (pictos) [inline XamarinCommunityToolkit/Views/CameraView/CameraView.shared.cs:16]: We can discuss this more and made changes in my PR... Since everything is inside the same assembly, we can change all these `EditorBrowsableState.Never` to `internal`.

I like this way because it's easy to hack, but I think the main idea to hide this is to protect developers from them selfs, so what do you think people?
A - Change to `internal`
B - Don't change at all.

Comment (pictos) [reply XamarinCommunityToolkit/Views/CameraView/CameraView.shared.cs:16]: Please mark people that should be part of this discussion

File: XamarinCommunityToolkit/Views/CameraView/CameraViewRenderer.ios.cs
Comment (pictos) [inline XamarinCommunityToolkit/Views/CameraView/CameraViewRenderer.ios.cs:112]: Another place that we use media source

File: XamarinCommunityToolkit/Views/CameraView/FormsCameraView.ios.cs
Comment (pictos) [inline XamarinCommunityToolkit/Views/CameraView/FormsCameraView.ios.cs:113]: This method isn't used, do we need it?

File: XamarinCommunityToolkit/Views/CameraView/FormsCameraView.ios.cs
Comment (pictos) [inline XamarinCommunityToolkit/Views/CameraView/FormsCameraView.ios.cs:13]: Just wondering if we need to make this class `sealed`. Since this is a `renderer` I think that we should allow the dev to extend this class. What do you think?

File: XamarinCommunityToolkit/Views/CameraView/FormsCameraView.ios.cs
Comment (pictos) [inline XamarinCommunityToolkit/Views/CameraView/FormsCameraView.ios.cs:234]: In the future, we want to allow the dev to save the media in another folder? I can see an enhancement issue for it in the future hehe

File: XamarinCommunityToolkit/Views/CameraView/MediaCapturedEventArgs.shared.cs
Comment (pictos) [inline XamarinCommunityToolkit/Views/CameraView/MediaCapturedEventArgs.shared.cs:6]: Since this is an EventArgs all properties should be read-only. What do you think? If yes I can change it on PR if needed

File: XamarinCommunityToolkitSample/Pages/Views/CameraViewPage.xaml.cs
Comment (pictos) [inline XamarinCommunityToolkitSample/Pages/Views/CameraViewPage.xaml.cs:10]: I love it!

File: XamarinCommunityToolkit/Views/CameraView/CameraView.shared.cs
Comment (jfversluis) [reply XamarinCommunityToolkit/Views/CameraView/CameraView.shared.cs:16]: Aaah right. Yeah, in Forms those are separate libraries. You're totally right. I hate doing these things. I was too much mindlessly copy/pasting these things ;) I'll fix that. Thanks

File: XamarinCommunityToolkit/Core/CameraMediaSource.shared.cs
Comment (jfversluis) [reply XamarinCommunityToolkit/Core/CameraMediaSource.shared.cs:11]: Talked about this on stream, seems not to be used right now, but let's keep it for the Android and UWP versions. If its still not used then, let's remove it. #notetoself

File: XamarinCommunityToolkit/Views/CameraView/MediaCapturedEventArgs.shared.cs
Comment (jfversluis) [reply XamarinCommunityToolkit/Views/CameraView/MediaCapturedEventArgs.shared.cs:6]: Changed this to have a internal constructor we can use, the rest is read-only. Good call!

File: XamarinCommunityToolkit/Core/StreamMediaSource.shared.cs
Comment (jfversluis) [reply XamarinCommunityToolkit/Core/StreamMediaSource.shared.cs:79]: Since it all works, I'm going to keep it like this for a first version. But we can definitely experiment with this.

Comment (jfversluis) [reply XamarinCommunityToolkit/Core/StreamMediaSource.shared.cs:79]: Also, still not sure this object is even used.

File: XamarinCommunityToolkit/Views/CameraView/CameraView.shared.cs
Comment (jfversluis) [reply XamarinCommunityToolkit/Views/CameraView/CameraView.shared.cs]: Changed these to be internal!

**Comment by bmacombe** [comment]
I think this is good to go also!

File: XamarinCommunityToolkit/Views/CameraView/CameraView.shared.cs
Comment (bmacombe) [reply XamarinCommunityToolkit/Views/CameraView/CameraView.shared.cs]: @AndreiMisiukevich @jfversluis UWP appears to use float vs double in it's camera control.  It's only the Zoom setting, so I can just cast back from double to float if you still think it best to keep double on CameraView.

**Comment by jfversluis** [comment]
Oh I see you had some more questions @pictos I didn't see them yesterday sorry about that. I'm going to merge this so I unblock you and we'll polish this later. Add the bits that you think should be changed about this in a PR so we can discuss it there.

**Comment by pictos** [comment]
@jfversluis don't worry, my last questions are minimal. It's ok to discuss that later (:

**Comment by acuntex** [comment]
There are still some bugs I noticed...
- if you put the app in background and then resume the app, the camera is frozen because the Camera got deactivated.
- Once you leave the page and use a BindingContext for the properties, all properties are reset and the whole RetrieveCameraDevice will run, which can slow down the closing process of the page closing.

**Comment by jfversluis** [comment]
Thanks for already testing this @acuntex ! I love the early feedback. 

Would you mind maybe opening separate issues for that with as much details as you have so we can track them? Thanks!

**Comment by acuntex** [comment]
Oh it's a pleasure, love it :-)

Wasn't sure if this repo is accepting bug reports, but I'll open them.

**Pull request merged**
